Tests/OpenCombineTests/PublisherTests/FlatMapTests.swift

```diff
@@ -0,0 +1,539 @@
+        // have one extra/buffered value after the downstream demand is satisfied.
+        // A demand of 1 on each child is created
+        XCTAssertEqual(childDemand, .max(1))
+
+        // Downstream demand is 2, so:
+        //  - this value gets sent
+        //  - downstream demand goes down to 1
+        //  - child is asked for 1 more
+        XCTAssertEqual(childPublisher.send(666), .max(1))
+        XCTAssertEqual(downstreamSubscriber.history, [.subscription("FlatMap"),
+                                                      .value(666)])
+
+        // Downstream demand is 1, so:
+        //  - this value gets sent
+        //  - downstream demand goes down to 0, but still need a buffered value
+        //  - child is asked for 1 more
+        XCTAssertEqual(childPublisher.send(777), .max(1))
+        XCTAssertEqual(downstreamSubscriber.history, [.subscription("FlatMap"),
+                                                      .value(666),
+                                                      .value(777)])
+
+        // Downstream demand is 0, so:
+        //  - this value is buffered and NOT sent
+        //  - downstream demand is 0 and there's a buffered value
+        //  - child is asked for 0 more
+        XCTAssertEqual(childPublisher.send(888), .max(0))
+        XCTAssertEqual(downstreamSubscriber.history, [.subscription("FlatMap"),
+                                                      .value(666),
+                                                      .value(777)])
+
+        childDemand = nil
+        downstreamSubscription!.request(.max(2))
+        XCTAssertEqual(childDemand, .max(1))
+        XCTAssertEqual(downstreamSubscriber.history, [.subscription("FlatMap"),
+                                                      .value(666),
+                                                      .value(777),
+                                                      .value(888)])
+    }
+
+    func testDemandFromUnlimitedToLimited() {
+        let upstreamPublisher = PassthroughSubject<Void, Never>()
+
+        var childDemand: Subscribers.Demand?
+        let childSubscription = CustomSubscription(onRequest: { childDemand = $0 })
+        let childPublisher = CustomPublisherBase<Int, Never>(
+            subscription: childSubscription)
+
+        let flatMap = upstreamPublisher.flatMap { _ in childPublisher }
+
+        var downstreamSubscription: Subscription?
+        let downstreamSubscriber = TrackingSubscriberBase<Int, Never>(receiveSubscription:
+        {
+            downstreamSubscription = $0
+            $0.request(.unlimited)
+        })
+
+        flatMap.subscribe(downstreamSubscriber)
+
+        upstreamPublisher.send(())
+
+        XCTAssertEqual(childDemand, .unlimited)
+
+        downstreamSubscription?.request(.max(3))
+        // TODO: Strangely Apple doesn't reset the child demand down to 1. That totally
+        // seems like a bug
+//        XCTAssertEqual(childDemand, .max(1))
```

Comment:
I don't think this is a bug.

As we've discovered, demand is additive, which means that if you request `.unlimited` and after that you request `.max(42)`, it should be completely the same as just requesting `.unlimited` once. If you blow up a dam, you can't stop the flow amymore :)

I believe this is what's happening here.